File: android/src/main/java/com/amplitude/skylab/AsyncFuture.java
package com.amplitude.skylab;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

class AsyncFuture<T> implements Future<T> {

    @Nullable private volatile T value = null;
    private volatile boolean completed = false;
    @Nullable private volatile Throwable throwable = null;
    @NotNull private final Object lock = new Object();

    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        return false;
    }

    @Override
    public boolean isCancelled() {
        return false;
    }

    @Override
    public boolean isDone() {
        return completed;
    }

    @Override
    @NotNull
    public T get() throws InterruptedException, ExecutionException {
        synchronized (lock) {
            while (!completed) {
                lock.wait();
            }
        }
        if (throwable != null) {
            throw new ExecutionException(throwable);
        }
        return value;
    }

    @Override
    @NotNull
    public T get(long timeout, @NotNull TimeUnit unit) throws InterruptedException, ExecutionException,
            TimeoutException {
        long nanosRemaining = unit.toNanos(timeout);
        long end = System.nanoTime() + nanosRemaining;
        synchronized (lock) {
            while (!completed && nanosRemaining > 0) {
                TimeUnit.NANOSECONDS.timedWait(lock, nanosRemaining);
                nanosRemaining = end - System.nanoTime();
            }
        }
        if (!completed) {
            throw new TimeoutException();
        }

        if (throwable != null) {
            throw new ExecutionException(throwable);
        }

        return value;
    }

    synchronized void complete(@NotNull T value) {
        if (!completed) {
            this.value = value;
            synchronized (lock) {
                completed = true;
                lock.notifyAll();
            }
        }
    }

    synchronized void completeExceptionally(@NotNull Throwable ex) {
        if (!completed) {
            throwable = ex;
            synchronized (lock) {
                completed = true;
                lock.notifyAll();
            }
        }
    }
}

File: sdk/src/main/java/com/amplitude/experiment/util/AsyncFuture.kt
package com.amplitude.experiment.util

import okhttp3.Call
import java.lang.NullPointerException
import java.util.concurrent.ExecutionException
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

internal class AsyncFuture<T>(private val call: Call? = null): Future<T> {

    @Volatile private var value: T? = null
    @Volatile private var completed = false
    @Volatile private var throwable: Throwable? = null
    private val lock = Object()

    override fun cancel(mayInterruptIfRunning: Boolean): Boolean {
        call?.cancel()
        return true
    }

    override fun isCancelled(): Boolean {
        return call?.isCanceled() ?: false
    }

    override fun isDone(): Boolean {
        return completed
    }

    @Throws(InterruptedException::class, ExecutionException::class)
    override fun get(): T {
        synchronized(lock) {
            while (!completed) {
                lock.wait()
            }
        }
        if (throwable != null) {
            throw ExecutionException(throwable)
        }
        return value
            ?: throw ExecutionException(NullPointerException("Future value must not be null"))
    }

    @Throws(InterruptedException::class, ExecutionException::class, TimeoutException::class)
    override fun get(timeout: Long, unit: TimeUnit): T {
        var nanosRemaining = unit.toNanos(timeout)
        val end = System.nanoTime() + nanosRemaining
        synchronized(lock) {
            while (!completed && nanosRemaining > 0) {
                TimeUnit.NANOSECONDS.timedWait(lock, nanosRemaining)
                nanosRemaining = end - System.nanoTime()
            }
        }
        if (!completed) {
            throw TimeoutException()
        }
        if (throwable != null) {
            throw ExecutionException(throwable)
        }
        return value
            ?: throw ExecutionException(NullPointerException("Future value must not be null"))
    }

    @Synchronized
    internal fun complete(value: T) {
        if (!completed) {
            this.value = value
            synchronized(lock) {
                completed = true
                lock.notifyAll()
            }
        }
    }

    @Synchronized
    internal fun completeExceptionally(ex: Throwable) {
        if (!completed) {
            throwable = ex
            synchronized(lock) {
                completed = true
                lock.notifyAll()
            }
        }
    }
}